codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,43 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
```

Comment:
I don't think this quite the same thing as state management.  WRT the flow controller, it just has to make sure that HEADERS are sent immediately following the last DATA frame sent for that stream.  If it's a new stream, no DATA will have been sent, so we can send the headers immediately.